File: litedb-java/src/main/java/com/litedb/cluster/ShardReplica.java
package com.litedb.cluster;

import java.io.IOException;
import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.Semaphore;

/**
 * ShardReplica — one node's replica of one shard: a {@link RaftGroup} bound to a {@link ShardStore},
 * plus the leader-side transactional commit path (conflict-check → assign HLC commit timestamp →
 * propose through Raft → wait for majority). For cross-shard transactions the 2PC coordinator drives
 * prepare/commit across several of these.
 *
 * The 2PC lock is a {@link Semaphore} (not a ReentrantLock) because prepare and commit arrive on
 * different RPC threads, and a semaphore permit may be released by a thread other than the acquirer.
 */
public final class ShardReplica {

    public final String nodeId;
    public final String shardId;
    private final Hlc hlc;
    public final ShardStore store;
    public final RaftGroup raft;
    private final Semaphore commitLock = new Semaphore(1, true);
    private final Map<String, Object[]> prepared = new ConcurrentHashMap<>();  // txnId -> [commitTs, writes]

    public ShardReplica(String nodeId, String shardId, List<String> peers, RaftGroup.Transport transport,
                        String dataDir, Hlc hlc, boolean preferred, RaftGroup.Events events) throws IOException {
        this.nodeId = nodeId;
        this.shardId = shardId;
        this.hlc = hlc;
        this.store = new ShardStore(dataDir + "/shard-" + shardId + "-data");
        this.raft = new RaftGroup(nodeId, shardId, peers, transport,
                (index, command) -> store.apply(command),
                dataDir + "/shard-" + shardId + "-raft", preferred, events);
    }

    public void start() { raft.start(); }
    public void stop() { raft.stop(); store.close(); }
    public boolean isLeader() { return raft.isLeader(); }
    public String leaderId() { return raft.leaderId(); }
    public long snapshotTs() { return store.snapshotTs(); }

    // ---- single-shard transactional write ---------------------------------

    public Map<String, Object> commitWrite(Map<String, Object> writes, Long readTs, long timeoutMs) {
        commitLock.acquireUninterruptibly();
        try {
            if (!raft.isLeader()) return err("not_leader", "leader", raft.leaderId());
            long rts = readTs != null ? readTs : store.snapshotTs();
            String conflict = checkConflicts(writes, rts);
            if (conflict != null) return err("conflict", "key", conflict);
            long commitTs = readTs != null ? hlc.update(rts) : hlc.now();
            Long index = raft.propose(buildCommand(commitTs, writes));
            if (index == null) return err("not_leader");
            boolean ok = raft.waitCommit(index, timeoutMs);
            if (!ok) return err("timeout");
            Map<String, Object> r = ok();
            r.put("commit_ts", commitTs);
            return r;
        } finally {
            commitLock.release();
        }
    }

    private String checkConflicts(Map<String, Object> writes, long readTs) {
        for (String key : writes.keySet()) {
            if (store.newestCommittedTs(key) > readTs) return key;
        }
        return null;
    }

    // ---- 2PC participant side ---------------------------------------------

    public Map<String, Object> prepare(String txnId, Map<String, Object> writes, long readTs, long commitTs) {
        commitLock.acquireUninterruptibly();
        boolean release = true;
        try {
            if (!raft.isLeader()) return err("not_leader", "leader", raft.leaderId());
            String conflict = checkConflicts(writes, readTs);
            if (conflict != null) return err("conflict", "key", conflict);
            prepared.put(txnId, new Object[]{commitTs, writes});
            release = false;  // hold the permit until commit/abort
            return ok();
        } finally {
            if (release) commitLock.release();
        }
    }

    @SuppressWarnings("unchecked")
    public Map<String, Object> commitPrepared(String txnId, long timeoutMs) {
        Object[] staged = prepared.remove(txnId);
        if (staged == null) return err("unknown_txn");
        try {
            long commitTs = (Long) staged[0];
            Map<String, Object> writes = (Map<String, Object>) staged[1];
            Long index = raft.propose(buildCommand(commitTs, writes));
            if (index == null) return err("not_leader");
            boolean ok = raft.waitCommit(index, timeoutMs);
            return ok ? ok() : err("timeout");
        } finally {
            commitLock.release();
        }
    }

    public Map<String, Object> abortPrepared(String txnId) {
        if (prepared.remove(txnId) != null) {
            commitLock.release();
        }
        return ok();
    }

    // ---- reads ------------------------------------------------------------

    public String read(String key, Long readTs) {
        return store.read(key, readTs != null ? readTs : store.snapshotTs());
    }

    public List<Map.Entry<String, String>> scan(String lo, String hi, Long readTs) {
        return store.scan(lo, hi, readTs != null ? readTs : store.snapshotTs());
    }

    public Map<String, Object> status() { return raft.status(); }

    // ---- helpers ----------------------------------------------------------

    private static Map<String, Object> buildCommand(long commitTs, Map<String, Object> writes) {
        List<Object> ws = new ArrayList<>();
        for (Map.Entry<String, Object> e : writes.entrySet()) {
            List<Object> kv = new ArrayList<>();
            kv.add(e.getKey());
            kv.add(e.getValue());  // null encodes a delete
            ws.add(kv);
        }
        Map<String, Object> cmd = new LinkedHashMap<>();
        cmd.put("ts", commitTs);
        cmd.put("writes", ws);
        return cmd;
    }

    private static Map<String, Object> ok() {
        Map<String, Object> m = new LinkedHashMap<>();
        m.put("ok", true);
        return m;
    }

    private static Map<String, Object> err(String error, Object... extra) {
        Map<String, Object> m = new LinkedHashMap<>();
        m.put("ok", false);
        m.put("error", error);
        for (int i = 0; i < extra.length; i += 2) m.put((String) extra[i], extra[i + 1]);
        return m;
    }
}

File: litedb-python/shard_replica.py
"""
shard_replica.py — one node's replica of one shard.

Binds a RaftGroup (consensus + replication) to a ShardStore (MVCC storage), and adds the
leader-side transactional commit path:

    conflict-check (OCC)  ->  assign HLC commit timestamp  ->  propose through Raft
                          ->  wait for majority commit + local apply

The whole single-shard commit is serialized by a per-shard lock so the conflict check sees every
prior committed write (first-committer-wins). For cross-shard transactions the 2PC coordinator
(txn_coordinator.py) drives prepare/commit across several of these.
"""

import os
import threading
from typing import Callable, Optional

from hlc import HLC
from raft_node import RaftGroup, SendFn
from shard_store import ShardStore


class ShardReplica:
    def __init__(self, node_id: str, shard_id: str, peers: list[str],
                 send_fn: SendFn, data_dir: str, hlc: HLC, preferred: bool = False,
                 on_event: Optional[Callable[[str, str], None]] = None):
        self.node_id = node_id
        self.shard_id = shard_id
        self.hlc = hlc
        self.store = ShardStore(os.path.join(data_dir, f"shard-{shard_id}-data"))
        self.raft = RaftGroup(
            node_id=node_id, group_id=shard_id, peers=peers, send_fn=send_fn,
            apply_fn=lambda index, command: self.store.apply(command),
            data_dir=os.path.join(data_dir, f"shard-{shard_id}-raft"),
            preferred=preferred, on_event=on_event,
        )
        # serialize the check+propose decision section per shard (the durable, cross-operation
        # "lock" on keys is the replicated intent itself, not this in-process lock).
        self._lock = threading.Lock()

    def start(self) -> None:
        self.raft.start()

    def stop(self) -> None:
        self.raft.stop()
        self.store.close()

    def is_leader(self) -> bool:
        return self.raft.is_leader()

    def is_ready(self) -> bool:
        return self.raft.is_ready()

    def leader_id(self) -> Optional[str]:
        return self.raft.leader_id

    def snapshot_ts(self) -> int:
        return self.store.snapshot_ts()

    # ---- single-shard transactional write --------------------------------

    def commit_write(self, writes: dict, read_ts: Optional[int] = None,
                     timeout: float = 3.0) -> dict:
        """writes: {user_key: value_or_None}. Returns {"ok": bool, ...}."""
        with self._lock:
            if not self.raft.is_leader():
                return {"ok": False, "error": "not_leader", "leader": self.raft.leader_id}
            if not self.raft.is_ready():
                return {"ok": False, "error": "not_ready"}
            rts = read_ts if read_ts is not None else self.store.snapshot_ts()
            conflict = self._check_conflicts(writes, rts, txn_id="")
            if conflict is not None:
                return {"ok": False, **conflict}
            # commit must be newer than the snapshot it was validated against
            commit_ts = self.hlc.update(rts) if read_ts is not None else self.hlc.now()
            index = self.raft.propose({"ts": commit_ts,
                                       "writes": [[k, v] for k, v in writes.items()]})
            if index is None:
                return {"ok": False, "error": "not_leader"}
            ok = self.raft.wait_commit(index, timeout=timeout)
            return {"ok": ok, "commit_ts": commit_ts} if ok else {"ok": False, "error": "timeout"}

    def _check_conflicts(self, writes: dict, read_ts: int, txn_id: str) -> Optional[dict]:
        """A write conflicts if a newer committed version exists (OCC) or a key is locked by another
        prepared 2PC intent."""
        for key in writes:
            if self.store.newest_committed_ts(key) > read_ts:
                return {"error": "conflict", "key": key}
            locker = self.store.intent_locking(key, exclude_txn=txn_id)
            if locker is not None:
                return {"error": "locked", "key": key, "by": locker}
        return None

    # ---- 2PC participant side (driven by the coordinator) ----------------

    def prepare(self, txn_id: str, writes: dict, read_ts: int, commit_ts: int,
                coordinator: Optional[str] = None) -> dict:
        """Phase 1: conflict-check, then replicate a PREPARE intent through Raft. Once the intent is
        committed to a majority it is durable AND survives a leadership change (any new leader has
        it in its log), so the vote can't be lost — the participant-recovery property."""
        with self._lock:
            if not self.raft.is_leader():
                return {"ok": False, "error": "not_leader", "leader": self.raft.leader_id}
            if not self.raft.is_ready():
                return {"ok": False, "error": "not_ready"}
            conflict = self._check_conflicts(writes, read_ts, txn_id=txn_id)
            if conflict is not None:
                return {"ok": False, **conflict}
            index = self.raft.propose({"op": "prepare", "txn_id": txn_id, "commit_ts": commit_ts,
                                       "writes": [[k, v] for k, v in writes.items()]})
            if index is None:
                return {"ok": False, "error": "not_leader"}
            ok = self.raft.wait_commit(index, timeout=3.0)  # hold lock until the intent is applied
            return {"ok": True} if ok else {"ok": False, "error": "timeout"}

    def commit_prepared(self, txn_id: str, timeout: float = 3.0) -> dict:
        """Phase 2 (commit): replicate a COMMIT entry; apply turns the intent's writes into committed
        versions. Idempotent — committing an already-resolved txn is a no-op on apply."""
        index = self.raft.propose({"op": "commit", "txn_id": txn_id})
        if index is None:
            return {"ok": False, "error": "not_leader"}
        ok = self.raft.wait_commit(index, timeout=timeout)
        return {"ok": True} if ok else {"ok": False, "error": "timeout"}

    def abort_prepared(self, txn_id: str, timeout: float = 3.0) -> dict:
        index = self.raft.propose({"op": "abort", "txn_id": txn_id})
        if index is None:
            return {"ok": False, "error": "not_leader"}
        ok = self.raft.wait_commit(index, timeout=timeout)
        return {"ok": True} if ok else {"ok": False, "error": "timeout"}

    # ---- reads ------------------------------------------------------------

    def read(self, key: str, read_ts: Optional[int] = None):
        rts = read_ts if read_ts is not None else self.store.snapshot_ts()
        return self.store.read(key, rts)

    def scan(self, lo: str, hi: str, read_ts: Optional[int] = None) -> list[tuple[str, str]]:
        rts = read_ts if read_ts is not None else self.store.snapshot_ts()
        return self.store.scan(lo, hi, rts)

    def status(self) -> dict:
        return self.raft.status()
